galaxyproject/galaxy
# Updated path-based interactive tools with entry point path injection, support for ITs with relative links, shortened URLs, doc and config updates including Podman job_conf

Work on path-based interactive tools started with issue #14690 concluded. Latest additions:

- Support for path based ITs with relative links, implemented through:
  - New entry point attribute `requires_path_in_url` to signal to gx-it-proxy whether the path should be stripped from the URL (if `False`, which is default) or not stripped (if `True`)
- Support for path based ITs with absolute links using one of:
  - Mechanism for injecting entry point URL path in HTTP header variable, implemented through:
    - New entry point attribute `label` to better identify specific entry point
    - New environment variable injection method named `entry_point_path_for_label`
  - Mechanism for injecting entry point URL path in environment variable at tool startup, implemented through:
    - New entry point attribute `requires_path_in_header named` to signal to gx-it-proxy that the entry point path should be added to the request as a HTTP header of the provided name (if set)
- All interactive tool URLs are shortened to the level of `interactivetools_shorten_url`, removing the need for this config (now removed). This has been implemented through:
  - Removing the unneeded `/access` part for path-based ITs
  - Identifying entry point class with `ep` instead of `interactivetoolentrypoint` (only one class implemented anyway)
  - Reducing entry point encoded identifier from 16 to 12-13 chars using base36 `[a-z][0-9]` instead of base16 `[a-f][0-9]` (hex)
  - Using 8-byte cryptographic token encoded with base36 (~13 chars) instead of hex-encoded uuid (32 chars)
  - `interactivetools_shorten_url` used hex-encoded entry point id (16 chars)+10 first chars of uuid = 26 chars. New base36-encoded identifier + token is 25-26 chars
  - `interactivetools_shorten_url` already stripped `.interactivetoolentrypoint`. Adding back `.ep` adds 3 chars but brings consistency. If shorter URLs are needed for the kubernetes use case, the galaxy.yml config `interactivetools_prefix="interactivetool"` can be shortened. @almahmoud looks ok?
  - Example domain-based URL: `https://24q1dbzrknq1v-1a1p13jnahscj.ep.interactivetool.myserver.net/`
  - Example path-based URL: `https://myserver.net/interactivetool/ep/24q1dbzrknq1v/1a1p13jnahscj/`
- Example tools for various modes have been properly configured:
  - Jupytool (`jupyter_notebook_1.0.0`) is now a path-based IT using entry point path injection into environment variable
  - OpenRefine and Ubuntu XFCE Desktop (`guacamole_desktop`) are now path-based ITs without path injection (as they only use relative links)
  - RStudio is now a path-based IT using entry point path injection as HTTP header
- Small updates of `galaxy.yml.sample`, `galaxy.yml.sample.interactivetools` and `tool_conf.xml.sample`
- Fetches `GALAXY_WEB_PORT` from `galaxy_infrastructure_web_port` instead of hardcoding to `8080` (all ITs)
- Reorganisation and update of interactive tools docs. Added section on path-based interactive tools. Updated sample configs
- Added `job_conf.yml.interactivetool`
- Added `job_conf.yml.interactivetool.podman` to document a working config to run ITs on Podman (@hexylena)
- Added note in `galaxy.yml.sample` on `interactivetools_map`
- Depends on `gravity` v1.0.4 (PR: https://github.com/galaxyproject/gravity/pull/114) , which again depends on `gx-it-proxy` version 0.0.6 (PR: https://github.com/galaxyproject/gx-it-proxy/pull/17)

Much details in a single PR here, but did not want to spend even more time on individual PRs.

FIxes #14690

Thanks to @morj-uio, @hexylena and @natefoo for help!

## How to test the changes?
(Select all options that apply)
- [ X ] This is a refactoring of components with existing test coverage.
- [ X  ] Instructions for manual testing are as follows:
  1. Open [https://proto2-test.fairtracks.net/](https://proto2-test.fairtracks.net/) 
  2. Start the updated interactive tools (Jupytool, RStudio, Open Refine, Ubuntu XFCE Desktop), notice that they are hosted under the same subdomain, but with ids  and tokens in the path.
  3. Notice that the tools still work.  

## License
